src/main.ts

```diff
@@ -167,31 +154,61 @@ export default class FolderNotesPlugin extends Plugin {
 
-		const excludedFolder = this.getExcludedFolderByPath(folder.path)
-		if (excludedFolder?.disableSync) return;
-		if (oldFileName !== folder.name + '.md') return;
-		let newFolderPath = this.getPathFromString(file.path);
-		if (newFolderPath.lastIndexOf('/') > 0) {
-			newFolderPath = newFolderPath.slice(0, newFolderPath.lastIndexOf('/')) + '/';
-		} else {
-			newFolderPath = '';
+		const newFilePath = this.getPathFromString(file.path);
+		const newFolder = this.app.vault.getAbstractFileByPath(newFilePath);
+
+		if(oldFolder) {
+			const excludedFolder = this.getExcludedFolderByPath(oldFolder.path)
+			if (excludedFolder?.disableSync) {
+				this.removeCSSClassFromTitleEL(oldFolder.path, 'has-folder-note')
+				this.removeCSSClassFromTitleEL(file.path, 'is-folder-note')
+				this.removeCSSClassFromTitleEL(oldPath, 'is-folder-note')
+			};
+			// nothing else for us to do as we have no new folder to action on
+			if (!newFolder) { return }
+
+			const newExcludedFolder = this.getExcludedFolderByPath(newFolder.path)
+			// Nothing else for us to do
+			if(newExcludedFolder?.disableSync || newExcludedFolder?.disableFolderNote) { return }
+		}
+
+		// file matched folder name before rename
+		// file hasnt moved just renamed
+		// Need to rename the folder
+		if (oldFolder && oldFolder.name + '.md' === oldFileName && newFilePath === oldFilePath) {
+			return this.renameFolderOnFileRename(file, oldPath, oldFolder)
+		}
+
+		// the note has been moved somewhere and is no longer a folder note
+		// cleanup css on the folder and note
+		if (oldFolder && oldFolder.name + '.md' === oldFileName && newFilePath !== oldFilePath) {
+			this.removeCSSClassFromTitleEL(oldFolder.path, 'has-folder-note')
+			this.removeCSSClassFromTitleEL(file.path, 'is-folder-note')
+			this.removeCSSClassFromTitleEL(oldPath, 'is-folder-note')
 		}
-		newFolderPath += file.name.replace('.md', '');
+
+		// file has been moved into position where it can be a folder note!
+		if(newFolder && newFolder.name === file.basename) {
+			this.addCSSClassToTitleEL(newFolder.path, 'has-folder-note')
+			this.addCSSClassToTitleEL(file.path, 'is-folder-note', true)
+		}
+	}
+
+	async renameFolderOnFileRename(file: TFile, oldPath: string, oldFolder: TAbstractFile) {
+		const newFolderPath = oldFolder.parent.path + '/' + file.basename;
 		if (this.app.vault.getAbstractFileByPath(newFolderPath)) {
-			this.app.vault.rename(file, oldPath);
+			await this.app.vault.rename(file, oldPath);
 			return new Notice('A folder with the same name already exists');
 		}
-		if (folder instanceof TFolder) {
-			let newPath = file.basename;
-			if (folder.path.indexOf('/') >= 0) {
-				newPath = this.getPathFromString(folder.path) + '/' + newPath;
-			}
-			this.app.vault.rename(folder, newPath);
-		}
+		await this.app.vault.rename(oldFolder, newFolderPath)
+	}
+
+	handleFileCreate(file: TFile) {
+		if(file.basename !== file.parent.name) { return }
```

Comment:
Can you add breakpoints to this function when creating notes to help track it down?